src/protocols/mcsp/MessageCounterManager.cpp

```diff
@@ -61,7 +61,82 @@ void MessageCounterSyncMgr::Shutdown()
     }
 }
 
-void MessageCounterSyncMgr::OnMessageReceived(Messaging::ExchangeContext * exchangeContext, const PacketHeader & packetHeader,
+MessageCounter & MessageCounterManager::GetGlobalUnsecureCounter()
+{
+    return mGlobalUnencryptedMessageCounter;
+}
+
+MessageCounter & MessageCounterManager::GetGlobalSecureCounter()
+{
+    return mGlobalEncryptedMessageCounter;
+}
+
+MessageCounter & MessageCounterManager::GetLocalSessionCounter(Transport::PeerConnectionState * state)
```

Comment:
Here, we are passing a `PeerConnectionState` into a `MessageCounterManager` instance to retrieve the `mSessionMessageCounter` from the passed `PeerConnectionState`.

If this pattern is retained, we should mark these getters static.

Larger question though: is this helper actually helpful?  This seems less intuitive than referencing `state->GetSessionMessageCounter().GetLocalMessageCounter()` directly.  The first thing I wondered looking at this PR is whether each `PeerConnectionState` had its own `mSessionMessageCounter`.  Indeed each one does.

But passing this through the `MessageCounterManager`, and particularly through a `MessageCounterManager` instance, makes this much less clear.

Reply:
The problem is that a dummy implementation `DummyMessageCounterManager` should be provided for test-cases and migration to by pass the counter sync and verification. So the function `GetLocalSessionCounter` need to be virtual.

Reply:
Well then this just seems like a very odd pattern to achieve exposure of a GetLocalSessionCounter method in `MessageCounterManagerInterface.h`.  In both the dummy implementation and the real implementation, this and related method bodies are the same: they just give back something from the passed `PeerConnectionState` pointer.

Why does `MessageCounterManagerInterface.h` need to expose things from `PeerConnectionState` in the first place?  In other words, why do we have an object interface definition that exposes something not contained within itself?

And if we are going to do this, can we at least mark these methods const?

Reply:
The problem comes form which instance holds the `SessionMessageCounter` object.

1. `SessionMessageCounter` object is bound to a session.
2. `MessageCounterManager` is the only class need to access the `SessionMessageCounter` object

So I'm force to put `SessionMessageCounter` inside the `PeerConnectionState` (due to 1), but I don't want to expose it to anything except the `MessageCounterManager` (due to 2). It is very hard to express using C++.

It looks a bit strange, but when we restrict that `PeerConnectionState::GetSessionMessageCounter` is only accessible by the `MessageCounterManager`, everything will make sense.

Reply:
With respect, it looks strange because it is strange.  Why is the singleton that holds the global secured and unsecured message counters also exposing the `SessionMessageCounter` interface?  They have completely different life cycles.  I don't see how entangling these provides value.

Seeing this, I had to check back and forth between the spec and the code to verify that the code actually does implement the spec correctly with the global unsecured and secured counter singletons and the per-session counters.  It turns out the code does do this, but that is significantly obfuscated by this construction.  Untangling this already cost engineering time in this review.  Why would we obligate all subsequent viewers of the code to repeat this exercise?

The case I would make is this: of course `SessionMessageCounter` lives inside `PeerConnectionState`.  _That's how this is supposed to work._  Please allow the code to reflect what it actually does.

With respect to this:

> 2\. `MessageCounterManager` is the only class need to access the `SessionMessageCounter` object
> 
> So I'm force to put `SessionMessageCounter` inside the `PeerConnectionState` (due to 1), but I don't want to expose it to anything except the `MessageCounterManager` (due to 2). It is very hard to express using C++.

But the PR does not do this.  `MessageCounterManager` is exposing the `SessionMessageCounter` to _other_ code in its public interface.  The case you're actually making is that the `MessageCounterManager` should provide the public interface for all counters.  I would instead suggest renaming `MessageCounterManager`; it's clearly very awkward for a single class to manage some things that last the lifetime of the program and other things that are instantiated on a per-session basis.

Reply:
I have moved them out of `MessageCounterManager` according to the suggestion. But I have to leave `IsSyncCompleted`, `VerifyCounter` and `CommitCounter` in the interface in order to implement the `DummyMessageCounterManager` interface.

Reply:
Thank you.